vending-app/src/main/kotlin/com/android/vending/installer/Install.kt

```diff
@@ -0,0 +1,231 @@
+                emitProgress(notifyId, InstallComplete)
+                        },
+            onFailure = { message ->
+                deferred.completeExceptionally(RuntimeException(message))
+                emitProgress(notifyId, InstallError(message ?: "UNKNOWN"))
+            }
+        )
+        val keyguardManager = getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
+        val intent = Intent(this, SessionResultReceiver::class.java)
+        intent.putStringArrayListExtra(SessionResultReceiver.KEY_TEMP_FILES, ArrayList(tempFiles))
+        intent.putExtra(SessionResultReceiver.KEY_NOTIFY_ID, notifyId)
+        val pendingIntent = PendingIntent.getBroadcast(
+            this, sessionId, intent,
+            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
+        )
+        emitProgress(notifyId, CommitingSession(createPendingIntent(VENDING_INSTALL_ACTION, sessionId, pendingIntent)
+            , createPendingIntent(VENDING_INSTALL_DELETE_ACTION, sessionId, null)))
+        if (!keyguardManager.isKeyguardLocked) {
+            session.commit(pendingIntent.intentSender)
+            // don't abandon if `finally` step is reached after this point
+            session = null
+
+            Log.d(TAG, "installPackages session commit")
+            return deferred.await()
+        } else {
+            Log.d(TAG, "installPackagesInternal: The screen is locked and waiting for the user to click the notification to install")
+            // don't abandon if `finally` step is reached after this point
+            session = null
+            throw Exception("The device screen is off, waiting for installation")
+        }
+    } catch (e: IOException) {
+        Log.w(TAG, "Error installing packages", e)
+        throw e
+    } finally {
+        // discard downloaded data
+        session?.let {
+            Log.d(TAG, "Discarding session after error")
+            it.abandon()
+        }
+    }
+}
+
+private fun Context.createInstallTempDir(packageName: String) : File {
+    val versionCode =PackageInfoCompat.getLongVersionCode(
+            packageManager.getPackageInfo(packageName, 0)
+    )
+
+    val tempDir = File(cacheDir, "temp_apk").apply {
+        if (!exists()) mkdirs()
+    }
+
+    val packageTempDir = File(tempDir, packageName).apply {
+        if (!exists()) mkdirs()
+    }
+
+    val versionTempDir = File(packageTempDir, versionCode.toString()).apply {
+        if (!exists()) mkdirs()
+    }
+    return versionTempDir
+}
+
+
+private fun createNotificationId(packageName: String, components: List<PackageComponent>) : Int{
+    val hash = (packageName + components.joinToString("") { it.componentName }).hashCode()
+    return hash and Int.MAX_VALUE
+}
+
+private fun Context.createPendingIntent(action: String, sessionId: Int, pendingIntent: PendingIntent? = null): PendingIntent {
+    val installIntent = Intent(this.applicationContext, InstallReceiver::class.java).apply {
```

Comment:
@DaVinci9196
It report: Class requires API level 21 (current min is 19): InstallReceiver [NewApi]
There should be an API check for this API.